versiondb/client/restore_app_db.go

```diff
@@ -0,0 +1,332 @@
+				// https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables
+				i := i
+				group.Submit(func() error {
+					return oneStore(stores[i], snapshots[i], iavlDir, sstFileSizeTarget, sorterChunkSize)
+				})
+			}
+
+			if err := group.Wait(); err != nil {
+				return errors.Wrap(err, "worker pool wait fail")
+			}
+
+			// collect the sst files
+			entries, err := os.ReadDir(iavlDir)
+			if err != nil {
+				return errors.Wrapf(err, "read directory fail: %s iavlDir")
+			}
+			sstFiles := make([]string, 0, len(entries))
+			for _, entry := range entries {
+				name := entry.Name()
+				if strings.HasPrefix(name, "tmp-") {
+					sstFiles = append(sstFiles, filepath.Join(iavlDir, name))
+				}
+			}
+
+			// sst files ingestion
+			ingestOpts := grocksdb.NewDefaultIngestExternalFileOptions()
+			defer ingestOpts.Destroy()
+			ingestOpts.SetMoveFiles(true)
+
+			db, err := grocksdb.OpenDb(open_db.NewRocksdbOptions(false), iavlDir)
+			if err != nil {
+				return errors.Wrap(err, "open iavl db fail")
+			}
+			defer db.Close()
+
+			if err := db.IngestExternalFile(sstFiles, ingestOpts); err != nil {
+				return errors.Wrap(err, "ingset sst files fail")
+			}
+
+			// write the metadata part separately, because it overlaps with the other sst files
+			if err := writeMetadata(db, &commitInfo); err != nil {
+				return errors.Wrap(err, "write metadata fail")
+			}
+
+			fmt.Printf("version: %d, app hash: %X\n", commitInfo.Version, commitInfo.Hash())
+			return nil
+		},
+	}
+
+	cmd.Flags().Uint64(flagSSTFileSize, DefaultSSTFileSize, "the target sst file size, note the actual file size may be larger because sst files must be split on different key names")
+	cmd.Flags().String(flagStores, "", "list of store names, default to the current store list in application")
+	cmd.Flags().Uint64(flagSorterChunkSize, DefaultSorterChunkSizeIAVL, "uncompressed chunk size for external sorter, it decides the peak ram usage, on disk it'll be snappy compressed")
+	cmd.Flags().Int(flagConcurrency, runtime.NumCPU(), "Number concurrent goroutines to parallelize the work")
+
+	return cmd
+}
+
+// oneStore process a single store, can run in parallel with other stores,
+func oneStore(store string, snapshot *memiavl.Snapshot, sstDir string, sstFileSizeTarget, sorterChunkSize uint64) error {
+	defer snapshot.Close()
+
+	prefix := []byte(fmt.Sprintf(storeKeyPrefix, store))
+
+	inputChan, outputChan := extsort.Spawn(sstDir, extsort.Options{
+		MaxChunkSize:      int64(sorterChunkSize),
+		LesserFunc:        compareSorterNode,
+		SnappyCompression: true,
+	}, PipelineBufferSize)
+
+	err := snapshot.ScanNodes(func(node memiavl.PersistedNode) error {
+		bz, err := encodeSorterNode(node)
+		if err != nil {
+			return err
+		}
+		inputChan <- bz
+		return nil
+	})
+	close(inputChan)
+	if err != nil {
+		return err
+	}
+
+	sstWriter := newIAVLSSTFileWriter()
+	defer sstWriter.Destroy()
+
+	sstSeq := 0
+	openNextFile := func() error {
```

Comment:
the caller should call the `Finish` on previous one before `openNextFile`.

Reply:
yea, should close already since openNextFile [after](https://github.com/crypto-org-chain/cronos/blob/499dff44915e0caad83243fbc546a42766ce119b/versiondb/client/restore_app_db.go#L216-L219) Finish